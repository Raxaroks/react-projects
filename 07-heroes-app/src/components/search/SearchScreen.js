import React, { useMemo } from 'react';
import queryString from 'query-string';     
import { useLocation } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import { getHeroesByName } from '../../selectors/getHeroesByName';
import HeroCard from '../heroes/HeroCard';

const SearchScreen = ({ history }) => {

    const location = useLocation();
    const { q = '' } = queryString.parse( location.search );

    const [ formValues, handleInputChange, ] = useForm( {
        searchInput: q
    } );

    const { searchInput } = formValues;

    const heroesFiltered = useMemo( () => getHeroesByName(q), [q] );

    const handleSearch = (evt) => {
        evt.preventDefault();
        history.push( `?q=${searchInput}` );
    }

    return (
        <div>
            <h1>Search Screen</h1> <hr/>

            <div className="row">
                <div className="col-5">
                    <h4>Search Form</h4>
                    <hr/>

                    <form onSubmit={ handleSearch }>
                        <input
                            type="text"
                            placeholder="Find your hero"
                            className="form-control"
                            name="searchInput"
                            autoComplete="off"
                            value= { searchInput }
                            onChange={ handleInputChange }
                        />

                        <button
                            type="submit"
                            className="btn m-1 btn-block btn-outline-primary"
                        >
                            Search...
                        </button>
                    </form>
                </div>

                <div className="col-7">
                    <h4>Results</h4>
                    <hr/>

                    {
                        ( q === '' ) // si el query está vacío
                            &&
                            <div className="alert alert-info">
                                Search for a hero
                            </div>
                    }

                    {
                        ( q !== '' && heroesFiltered.length === 0 ) // si el query tiene contenido pero no encuentra alguna coincidencia
                            &&
                            <div className="alert alert-danger">
                                There is no hero with { q }
                            </div>
                    }


                    {
                        heroesFiltered.map( hero => (
                            <HeroCard
                                key={ hero.id }
                                {...hero}
                            />
                        ) )
                    }
                </div>
            </div>
        </div>
    );
}

export default SearchScreen;