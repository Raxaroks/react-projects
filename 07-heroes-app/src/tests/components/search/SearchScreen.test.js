import { mount } from "enzyme";
import { MemoryRouter, Route } from "react-router-dom";
import SearchScreen from "../../../components/search/SearchScreen";

describe('Pruebas en el componente: SearchScreen', () => {

    test('Debe de renderizarse correctamente con sus valores por defecto.', () => {
        const componentWrapper= mount(
            <MemoryRouter initialEntries={ ['/search'] }>
                <Route path="/search" component={ SearchScreen } />
            </MemoryRouter>
        );

        expect(componentWrapper).toMatchSnapshot();
        expect( componentWrapper.find('.alert-info').text().trim() ).toBe('Search for a hero');
    });    

    test('Debe de mostrarse a Batman y el input del valor del QueryString de la búsqueda.', () => {
        const componentWrapper= mount(
            <MemoryRouter initialEntries={ ['/search?q=batman'] }>
                <Route path="/search" component={ SearchScreen } />
            </MemoryRouter>
        );

        expect( componentWrapper.find('input').prop('value') ).toBe('batman');
        expect(componentWrapper).toMatchSnapshot();
    });    

    test('Debe de mostrar un error si no se encuentra el héroe en la búsqueda.', () => {
        const componentWrapper= mount(
            <MemoryRouter initialEntries={ ['/search?q=batman12345'] }>
                <Route path="/search" component={ SearchScreen } />
            </MemoryRouter>
        );

        expect( componentWrapper.find('.alert-danger').text().trim() ).toBe( `There is no hero with batman12345` );
        expect(componentWrapper).toMatchSnapshot();
    });    

    test('Debe de llamar al push del history.', () => {
        const historyMock = {
            push: jest.fn(),
        };

        const componentWrapper= mount(
            <MemoryRouter initialEntries={ ['/search?q=batman12345'] }>
                <Route path="/search" component={ () => <SearchScreen history={ historyMock } /> } />
            </MemoryRouter>
        );

        componentWrapper.find('input').simulate('change', {
            target: {
                name: 'searchInput',
                value: 'batman'
            }
        });

        componentWrapper.find('form').prop('onSubmit')({
            preventDefault() {  }
        });

        expect( historyMock.push ).toHaveBeenCalledWith( '?q=batman' );
    });
    
});
