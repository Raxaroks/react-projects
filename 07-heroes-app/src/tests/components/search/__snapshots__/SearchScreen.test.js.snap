// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente: SearchScreen Debe de mostrar un error si no se encuentra el héroe en la búsqueda. 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
   
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchInput"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman12345"
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-danger"
      >
        There is no hero with 
        batman12345
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en el componente: SearchScreen Debe de mostrarse a Batman y el input del valor del QueryString de la búsqueda. 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
   
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchInput"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman"
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="card ms-3"
        sytle={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Batman"
              className="card-img"
              src="./assets/heroes/dc-batman.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Batman
              </h5>
              <p
                className="card-text"
              >
                Bruce Wayne
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  Detective Comics #27
                </small>
              </p>
              <a
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                Más...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en el componente: SearchScreen Debe de renderizarse correctamente con sus valores por defecto. 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
   
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchInput"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search for a hero
      </div>
    </div>
  </div>
</div>
`;
